codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -395,17 +396,25 @@ void onBytesConsumed(@SuppressWarnings("unused") ChannelHandlerContext ctx,
         READ_PROCESSED_OK_TO_PROCESS_MORE
     }
 
+    private boolean initialWritability(DefaultHttp2FrameStream stream) {
+        // If the stream id is not valid yet we will just mark the channel as writable as we will be notified
+        // about non-writability state as soon as the first Http2HeaderFrame is written (if needed).
+        // This should be good enough and simplify things a lot.
+        return !isStreamIdValid(stream.id()) || isWritable(stream);
```

Comment:
This may not be correct in all cases. The peer can send a settings frame with initial_window_size of 0 which would mean new streams are not writable. A server/client may do this if they are being overwhelmed and/or was overly optimistic about how many resources it initially advertised ... not sure how common it is in practice.

We should consider checking the peer's initial window size if this is a locally created stream.

Reply:
The writability needs to be true so the application provides the initial headers. After we have the headers the stream exists and this won't be used any more. So the application shouldn't end up writing any DATA.

Reply:
I think what @ejona86 said is correct and so it should be correct @Scottmitch

Reply:
Ahh I see ... yes this makes sense. Although we don't make the same consideration for trailers. We don't really know when the trailers will be written so it is possible the user will write data until not writable and then want to write trailers but if they are using channel writability to conditionally write then they will have to wait. Writing trailers maybe less common, but it seems strange to exhibit different behavior in these scenarios.

Reply:
This has nothing to do with HEADERS vs DATA and how they are impacted by HTTP/2 flow control. It has everything to do with "the stream doesn't exist before the initial headers." So I don't really see the trailers as being weird at all. If client could send multiple headers (say, like server can with 1xx codes) then we'd still want this same logic.

Reply:
Yip understood this is to facilitate stream creation.